### ISSUE TITLE
Add Debian/Ubuntu packaging

### ISSUE THREAD

**Issue by MOZGIII** [ahmetb/kubectx#17]
What do you think about building installation packages? I'm interested in Debian/Ubuntu in particular.

**Comment by ahmetb** [comment]
@MOZGIII I don’t use Linux and I am unfamiliar with deb/rpm packaging. I am open to accepting contributions if you’re willing to work on one and get it reviewed by someone else.

**Comment by fiws** [comment]
Something like [fpm](https://github.com/jordansissel/fpm) could help here.

**Comment by AlienResidents** [comment]
I think the easiest way forward for this type of request might be to implement a package for https://snapcraft.io/ (I hope external l inks are ok in this comment).  It's very similar to how `brew` works for Macs.

I'm not sure of the details of creating a `snap` package, but I'm a willing learner if it seems like an interest for this project (All those Linux distros not using DEBs, or RPMs can still use a snap package).

**Comment by ahmetb** [comment]
@AlienResidents I won’t stop you from trying it. 

We are planning a rewrite in Go #115 which will allow us to use goreleaser. Goreleaser already can:
- create Snapcraft packages 
https://goreleaser.com/customization/#Snapcraft
- deb/rpm packages via NFPM (a fpm inspired tool) 
https://goreleaser.com/customization/#NFPM

So if you just wait on the rewrite in Go, we’ll already get to  both of the proposed solutions.

**Comment by AlienResidents** [comment]
> So if you just wait on the rewrite in Go, we’ll already get to both of the proposed solutions.

Wait I will, and thanks for the info! (:

**Comment by stormmore** [comment]
I am a fan of snaps and see that a few K8s tools now have those as an option include `kubectl` and `helm` which is awesome. Good to see that it will be an option here once a rewrite is complete!

**Comment by nsteinmetz** [comment]
Hi there,

For those interested, I have a first version of kubectx package for fedora. Still work in progress but it progress (just installing the two binaries so far, need to add other files soon).

No repo yet but I can share spec files, srpms and rpms for those interested.

Should I create a dedicated issue for rpm based packaing or do we gather all here ?

Let me know !

**Comment by MOZGIII** [comment]
Feel free to create a separate issue. Since Debian and Ubuntu are using apt - and not rpm - it won't really solve the the original issue, and it makes sense to track it separately.

**Comment by ahmetb** [comment]
Beware that there's a proposal (although hasn't started yet) to rewrite kubectx in Go (#115). If it happens, we'll probably use Goreleaser's support for [nfpm](https://github.com/goreleaser/nfpm) to easily create packages for these platforms.

But I still have no idea how these can possibly get into the centralized package repositories for those distros.

**Comment by cablespaghetti** [comment]
I'm in the process of building a snap for the new Go release at the moment: https://github.com/cablespaghetti/kubectx-snap

When I've got it all working including auto-completion I'll have to submit it to the snapcraft team to allow it to use "classic" confinement. Otherwise I don't believe there is a way to access hidden files within a user's home directory (such as .kube/config).

Is this something you want to manage within this repo @ahmetb or should I manage it as a separate repository and build some automation to update it on new releases of this tool?

Thanks

**Comment by ahmetb** [comment]
@cablespaghetti let's keep it out for now, then we can probably move depending on the situation. right now many other packagers (debian, alpine) live outside this repo.

**Comment by cablespaghetti** [comment]
I've got the snap including autocompletion working. Unfortunately someone has reserved the `kubectx` name on snapcraft.io but not published anything so I've filed a dispute and hopefully it will be released and I can move on to getting approval for classic confinement and finally get it published.

**Comment by ahmetb** [comment]
@cablespaghetti thanks. Make sure to package only bash-based implementations for now. Go version is not ready for prime-time.

**Comment by cablespaghetti** [comment]
Ok. I've packaged the Go version but can mark that as beta rather than stable for now if that's preferred. Will look at Bash tomorrow.

**Comment by ahmetb** [comment]
I recommend packaging bash for O(months). I’ll be reaching out to all packagers once Go version is ready for prime-time.

**Comment by cablespaghetti** [comment]
Ok makes sense. I've changed my master branch to be bash and will have a separate branch for Go which can be in the "beta" channel for the snap. Going to take a while before it gets published but will post here when it is.

**Comment by sstubbs** [comment]
A snap of this would be great.

**Comment by ahmetb** [comment]
It seems this is now packaged in many debian/ubuntu repos.

[![Packaging status](https://repology.org/badge/vertical-allrepos/kubectx.svg)](https://repology.org/project/kubectx/versions)

**Comment by mhalano** [comment]
I could check and Ubuntu will get too soon (for 21.10). My main concern is the Debian/Ubuntu version is the Bash one, not the Go one.

**Comment by mhalano** [comment]
@cablespaghetti Would you be able to publish the Snap version?

**Comment by cablespaghetti** [comment]
Hi. Unfortunately I didn't have any luck convincing the Snapcraft people that this snap should use "classic" confinement and I couldn't work out a way to make it usable otherwise. I am happy to assist with transferring of the kubectx registered snap name to someone else who is able to take this futher.

I am currently using https://github.com/devops-works/binenv to install kubectx.

**Comment by tunix** [comment]
> Hi. Unfortunately I didn't have any luck convincing the Snapcraft people that this snap should use "classic" confinement and I couldn't work out a way to make it usable otherwise. I am happy to assist with transferring of the kubectx registered snap name to someone else who is able to take this futher.
> 
> I am currently using https://github.com/devops-works/binenv to install kubectx.

Are there any other forum posts besides [this one](https://forum.snapcraft.io/t/request-for-classic-confinement-kubectx/17554/2)? I can see that the snapcraft team has responded your initial post with reasonable alternatives. Have you tried their suggestion of using `personal-files` interface?

**Comment by tunix** [comment]
@ahmetb - are we ready for packaging the go version as snap as we have an announcement in `README.md` ?

**Comment by ahmetb** [comment]
The Go version still has some shortcomings and doesn't work as well as the bash version which shells out to kubectl yet. I would still recommend keeping the debian/ubuntu packages with bash implementation.

**Comment by mhalano** [comment]
Hi! I will try next weekend (busy week ahead) to package kubectx+kubens as Snap.

**Comment by mhalano** [comment]
@cablespaghetti Can you deliver to me the 'kubectx' name on Snap store? I created a functional Snap without --classic confinment. It uses a special interface released in the newer versions of snapd. I made a PR to the project, I just need the name to publish and do some burocratic stuff around alias and auto-connection.

**Comment by cablespaghetti** [comment]
I'll message you directly, as there is a process I need to follow to do this 👍
